### ISSUE TITLE
New item

### ISSUE THREAD

**Issue by mmarcotte** [ustaxcourt/ef-cms#7932]
## Pre-Conditions

- Review spike: https://github.com/flexion/ef-cms/issues/8073
 - ustaxcourt/ef-cms-flexion#8508, ustaxcourt/ef-cms#7931

## Acceptance Criteria

- As a Tech Lead, in order to reduce human error, automate the process of switching colors after re-indexing is complete and [smoketests  have passed](https://github.com/flexion/ef-cms/issues/8509)
- As a developer deploying to prod, so that I understand the updated deployment process, I need the blue-green-migration.md document to be up to date with the latest changes made by this adjustment.

## IRS API Considerations
Do these changes impact the IRS API?

## Security Considerations
 - [ ] Does this work make you nervous about privacy or security?
 - [ ] Does this work make major changes to the system?
 - [ ] Does this work implement new authentication or security controls?
 - [ ] Does this work create new methods of authentication, modify existing security controls, or explicitly implement any security or privacy features?

## Notes

- Determining when re-indexing is complete is going to be hard. 

## Tasks
- [x] Feedback - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Hold reindex step pending approval in circleCI - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Create `reindex_status_lambda` - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Add role for reindex_status - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Add terraform infrastructure to automate cron to trigger `reindex_status_lambda` every 10 min - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Enable cron at the last step of migrate in circleCI and disable cron by default - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Once reindexing is complete, approve pending job in circleCI to continue - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
  - [x] Using the circle_workflow_id, get all jobs in the workflow, then with that grab the approval_request_id (which is the pending), then send command to approve job - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
  - [x] To approve a job, must have a personal_token so create a machine_user per https://support.circleci.com/hc/en-us/articles/360060360811-CircleCI-API-v2-returns-404-Not-Found-#:~:text=token%20instead[…]-,personal,-API%20token
  - [x] Update handler to approve job when reindexing has completed - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] test running through pipeline without a migration  https://app.circleci.com/pipelines/github/flexion/ef-cms/30981/workflows/77d315a3-c74f-4bf4-8897-3cce403e51b0 - Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] test running through pipeline with a migration Rosie, Cody, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] use machine_user token in the ENV of circle project settings Rosie, Mrinalini, Rae ustaxcourt/ef-cms-flexion#8973
- [x] Disable cron after the pending job has been approved Rosie, Mrinalini ustaxcourt/ef-cms-flexion#8973
- [x] DOD
  - [x] test reindex-status.js - Eric
  - [x] test `check-reindex-complete` Rosie, Mrinalini, Rae, Cody ustaxcourt/ef-cms-flexion#8973

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment

**UX**
 - [ ] Business test scenarios to meet all acceptance criteria have been written
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] Deployed to any Flexion environment for UX Review
 - [x] Deployed to the Court's migration environment

**Comment by codyseibert** [comment]
We tested this code on the court's dev on 8/31/2021 and it appeared to wait until everything was reindexed as expected.

**Comment by mariahkannenberg** [comment]
@cholly75 This is ready for your review.

**Comment by mmarcotte** [comment]
So close, but I think it's also too close for comfort. [The CircleCI job](https://app.circleci.com/pipelines/github/ustaxcourt/ef-cms/2734/workflows/25ed4695-d1bd-4c94-993e-98972a3eccc4/jobs/26695) timed out at 5 hours when it was very close to completion. 

The bones of this solution are excellent, and I think we can and should make use of them. It's just that the long running items with the deployment steps will need to happen outside of Circle. 

I think we need to use CircleCI for the testing and Continuous Integration, but I think the deploy automation might need to happen via Lambda, crons, and then call Circle's API in order to perform the Smoketests step and complete the automation.

**Comment by mariahkannenberg** [comment]
@codyseibert See above comments

**Comment by codyseibert** [comment]
doing some research, there may be a way to "block" a build to wait for approval.  We could use a lambda cron to invoke the circle job's Approve a job to have it continue with the reindexing is done.  https://circleci.com/docs/api/v2/#operation/getWorkflowById

**Comment by mmarcotte** [comment]
As this is where most of time takes place on the CircleCI job, I think this is a viable path forward.

**Comment by mariahkannenberg** [comment]
There's more work to do on this - not meeting AC.

**Comment by mariahkannenberg** [comment]
Added Ready for Engineering Tag - This story in on hold while Cody joins another mob to help w/onboarding. Rae said she could pick up this story when 8896 is done.

**Comment by mmarcotte** [comment]
https://app.circleci.com/pipelines/github/ustaxcourt/ef-cms/2780/workflows/d9c54ece-d6e8-4c8a-96ae-4c2038b07b83 

Great work team! 

@ericsorenson -- Just want to be sure that any specific steps that need to be run for this to work in Prod are documented in the PR for the Batch.

**Comment by mariahkannenberg** [comment]
@mmarcotte, happy to provide this documentation! Moving forward, can you add as part of the acceptance criteria of the story please :pray::skin-tone-3: Thank you!  For context, it's just much easier to have everything we need to do in one place. Otherwise, we may forget.

**Comment by mmarcotte** [comment]
@mariahkannenberg -- I think this should be automatic AC / DoD for every User Story. It would seem redundant to add it in one by one since it is vital whenever this information is required. And we might not know when we draft a story whether or not special instructions will have to be executed upon deployment as that is part of the solution.

In this story, there is AC for updating the blue-green deployment documentation. However, this story also has extra steps in order to deploy it properly the first time.
